### PR TITLE
[Filebeat] Improve ASA/FTD pipeline

### LINKED ISSUE #21658: Filebeat Cisco module doesn't correctly parse 305011, 302015, 302013, or 722051 message types
Version: 7.9.2
Platform: Debian 10.5
cf. [https://discuss.elastic.co/t/filebeat-cisco-module-cant-parse-305011-302015-302013-or-722015-message-types/250882](url)
Related issue: [https://github.com/elastic/beats/pull/17964](url)

Summary: dissect patterns in asa-ftd-pipeline.yml do not parse Cisco message numbers 305011, 302015, or 302013 when a user identity is present.

Recommended Solution: replace dissect patterns with grok patterns if multiple variations are required.
Note that the Logstash Cisco grok patterns for the 305011 and 302... messages correctly parse them.

The Cisco documentation makes it clear that the user name may be present in the log messages for all three of these message types.  [https://www.cisco.com/c/en/us/td/docs/security/asa/syslog/b_syslog/syslog-messages-302003-to-342008.html](url)

Sample messages:
%ASA-6-305011:
`Built dynamic UDP translation from any:10.10.142.244/62776(LOCAL\username) to USER-VLAN1120:10.11.121.66/62776
 Built dynamic TCP translation from any:10.10.128.177/50350 to VLAN1120:10.11.121.118/50350`

%ASA-6-302013:
`Built inbound TCP connection 2568541635 for VLAN1120:10.10.192.226/52232 (10.11.121.28/52232)(LOCAL\username1) to VLAN1120:172.217.4.46/443 (172.217.4.46/443) (username1)`
`Built inbound TCP connection 2568541636 for VLAN1120:10.10.142.144/54095 (10.11.121.70/54095) to VLAN1120:172.217.4.46/443 (172.217.4.46/443) (username2)`
`Built outbound TCP connection 2568541648 for VLAN1120:10.10.138.94/54582 (10.10.138.94/54582) to identity:172.217.4.46/443 (172.217.4.46/443)`

%ASA-6-302015:
`Built inbound UDP connection 2568541626 for VLAN1120:10.10.142.244/62776 (10.11.121.66/62776)(LOCAL\username) to VLAN1120:172.217.4.46/443 (172.217.4.46/443) (username)`

Furthermore, message 722051 doesn't parse correctly when there are angle brackets separating field values in the text, e.g. 
`Group <group-policy> User <user1> IP <10.10.10.10> IPv4 Address <10.11.11.11> IPv6 address <::> assigned to session`

### PR DESCRIPTION
**Pull request by hitchatwork**
## What does this PR do?

Fixes #21658

For messages 716002:
-Changed to GROK; allows for better parsing of event.reason
-Added field for cisco.webvpn.group.name
-Added field for event.reason per cisco docs for why session was terminated

For messages 722051:
-Add angle brackets to dissect to properly dissect the message, per cisco docs
-Added field for cisco.webvpn.group.name

For messages 305011:
-Change to GROK; allows for variance in message format with identity firewall

For messages 302013/302015:
-Change to GROK; allows for variance in message format with identity firewall

For messages 302014/302016/302021:
-Added patterns and modified order of patterns of GROK to better match teardown messages
-Note that order of processing is important as the most specific messages are matched first, falling through to the appropriate match.
-Added temp fields for teardown initiator and user; defined in cisco docs but currently no real place to put them, but could be in future.
-Added icmp.request.type and icmp.request.code parsing for 302021 messages
-Changed typecast to 'long' for network.bytes; it is poosible this value exceeds integer range, and ECS reference field is defined as long. This prevents processor failures (spotted in live data set when value is greater than int; changing typecast fixed).
-Changed duration matching from TIME to NOTSPACE, as long-lived connections (over 24 hours) don't match TIME.

Additionally, added descriptions to many fields to make them easier to find in 7.9 Kibana Ingest Node Pipeline editor.

## Why is it important?

Improved parsing of ASA/FTD messages mentioned above.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ]

## How to test this PR locally

Test against ASA and FTD logs; can be provided if needed.

## Related issues

- Closes #21658

### CODE REVIEW

**Comment by hitchatwork** [comment]
After latest CI build attempt, the event.reason field is being flagged as not previously defined. With the enhanced grok'ing, this is useful, especially to SIEM investigators.

Also, added fields for icmp types [again, flagged as previously not defined] when present that match up with packetbeat's definitions, which also can be useful to SIEM users.

**Comment by andrewstucki** [comment]
Hi @hitchatwork thank you for the contribution, I haven't yet taken a look at the actual pipeline changes yet, but from a quick glance at the build in Travis, you're hitting a number of

```
Exception: Key '*' found in event is not documented!
```

What those are pointing to is that you'll need to add the specific additional fields that are now filled in the corresponding `fields.yml` files that are used to generate the Elasticsearch index mapping. For the cisco `asa` module that should be under `x-pack/filebeat/module/cisco/asa/_meta/fields.yml`

**Comment by hitchatwork** [comment]
Hi @andrewstucki - thanks for the pointer on updating those files.  I updated the ASA and FTD meta fields.yml files, but now travis has thrown:
"Error: some files are not up-to-date. Run 'mage fmt update' then review and commit the changes. Modified: [filebeat/docs/fields.asciidoc]"

Not sure what to do here.  Any advice appreciated.

**Comment by P1llus** [comment]
@hitchatwork if you run "mage update" in `./beats/x-pack/filebeat` and then again in `./beats/filebeat` you should see some new files being changed from the output of `git status`, please add those as well.

It complains that when you update fields.yml, the updates do not exist in the overall field definition for filebeat, which is done through mage update.

Hope that helps! :)

**Comment by hitchatwork** [comment]
> @hitchatwork if you run "mage update" in `./beats/x-pack/filebeat` and then again in `./beats/filebeat` you should see some new files being changed from the output of `git status`, please add those as well.
> 
> It complains that when you update fields.yml, the updates do not exist in the overall field definition for filebeat, which is done through mage update.
> 
> Hope that helps! :)

Mage keeps complaining about files not found, and it appears the magefile references a 'v7' folder structure that isn't present:
Error parsing magefiles: error running "go list -f {{.Dir}}||{{.Name}} github.com/elastic/beats/v7/filebeat/scripts/mage/generate": exit status 1
can't load package: package github.com/elastic/beats/v7/filebeat/scripts/mage/generate: cannot find package "github.com/elastic/beats/v7/filebeat/scripts/mage/generate" in any of:

I checked on github and in my repositories and I don't see a v7 anywhere.

Thoughts?

**Comment by kvch** [comment]
`v7` is part of the import path of Beats, as we are currently in the v7 major release. It is not reflected in the folder structure.

What is in your `GOROOT`, `GOPATH` env variable and `go.mod` file?

**Comment by hitchatwork** [comment]
@kvch Appreciate your help here.  Answers to your questions:

I'm using Ubuntu Server 18.04

GOPATH="/home/hitch/"
GOROOT="/usr/lib/go-1.10"

The binary is at /usr/bin/go

I cloned the beats repository to:
/home/hitch/src/github.com/elastic/beats

For go.mod:
hitch@server:~/src/github.com/magefile/mage$ cat go.mod 
module github.com/magefile/mage

go 1.12
hitch@server:~/src/github.com/magefile/mage$

**Comment by hitchatwork** [comment]
@kvch So I had an outdated version of go that was causing my initial problems.

Now I'm able to run mage update, but when I do that in the beats/filebeat directory, it throws this error at the end:
```
warn: failed to upgrade pip (ignoring): failed to run "/home/hitch/src/github.com/hitchatwork/beats/build/ve/linux/bin/pip install -U pip: fork/exec /home/hitch/src/github.com/hitchatwork/beats/build/ve/linux/bin/pip: no such file or directory"Error: failed to run "/home/hitch/src/github.com/hitchatwork/beats/build/ve/linux/bin/pip install -U wheel: fork/exec /home/hitch/src/github.com/hitchatwork/beats/build/ve/linux/bin/pip: no such file or directory"
hitch@elk-02:~/src/github.com/hitchatwork/beats/filebeat$
```

Go env:
```
GOPATH="/home/hitch/go"
GOROOT="/usr/local/go"
```

I also searched a bit, and tried deleting the build/ve directory.  But same error.  I checked to make sure I have python3 and the venv module, which I do:
```
hitch@elk-02:~/src/github.com/hitchatwork/beats/filebeat$ python3
Python 3.6.9 (default, Oct  8 2020, 12:12:24) 
[GCC 8.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import venv
>>> 
```
Also, I am running Ubuntu Server 18.04

**Comment by jsoriano** [comment]
@hitchatwork do you have the `python3-venv`, or `python3.6-venv`, packages installed? `import venv` is not enough check because this always work in python 3, this package includes some scripts that are needed to actually create virtual environments.

If you have it, try to remove the virtual environments (the `/home/hitch/src/github.com/hitchatwork/beats/build/ve` directory) and run `mage update` again.

**Comment by kvch** [comment]
jenkins run tests

**Comment by hitchatwork** [comment]
@kvch and @jsoriano - thanks for the pointers.  Was finally able to get all the dependencies/env setup.

Now the only issues I have from the Travis build are:
1) Dictionary items added or removed (not sure how to fix this)
2) Grok patterns not matching but I believe the test files have invalid tests patterns (I opened issues #22042 and #22056 that identify what I believe the problems with the test files are)
3) Dictionary items changed - where the username contained the open and closing parentheses; I cleaned this up with the grok expressions since the username shouldn't actually contain the parentheses anyways.

Thanks!

**Comment by hitchatwork** [comment]
Swill awaiting a reviewer and/or tips fixing dictionary keys mentioned in my last comment/question.

Thanks!
👍

**Comment by jsoriano** [comment]
@hitchatwork oops, sorry for the long delay, could you please update your branch with master before continuing? Thanks!

**Comment by hitchatwork** [comment]
> @hitchatwork oops, sorry for the long delay, could you please update your branch with master before continuing? Thanks!

Updated branch with master.

**Comment by jsoriano** [comment]
jenkins run the tests please

**Comment by jsoriano** [comment]
Linting issue doesn't seem related now, let's see if https://github.com/elastic/beats/pull/23313 fixes this...

**Comment by jsoriano** [comment]
@hitchatwork sorry, could you update the branch with master again?

**Comment by hitchatwork** [comment]
@jsoriano no worries - just updated again.

**Comment by jsoriano** [comment]
jenkins run the tests again please

**Comment by jsoriano** [comment]
@hitchatwork I think that the pending errors are caused by your changes in the pipeline. This is not necessarily bad :slightly_smiling_face: but you will need to regenerate the rest of test files in the module.

For that, run the following command from the `x-pack/filebeat` directory:
```
GENERATE=1 TESTING_FILEBEAT_MODULES=cisco mage pythonIntegTest
```
(it can take a while)

If after that you still have errors, you may need to review your changes in the pipeline.

You can also try to fix the IP addresses you reported in https://github.com/elastic/beats/issues/22042, and regenerate with the previous command.

@adriansr do you know if these 5-octet "IPs" reported in https://github.com/elastic/beats/issues/22042 are correct? could they be changed to normal IPs or does this fifth number represent some other thing?

**Comment by hitchatwork** [comment]
@jsoriano Thanks for that pointer.  All but the two tests with the funky IP address passed.

I did fix those log message lines and posted a PR #23331 (but that will need you or someone's review there too).  Might be worth reviewing/posting that first, and then kicking this one off again (since the test pulls the sample logs from the elastic repo and not mine).

**Comment by jsoriano** [comment]
jenkins run the tests please

**Comment by jsoriano** [comment]
@hitchatwork #23331 has been merged, could you update this branch?

**Comment by hitchatwork** [comment]
@jsoriano can you kick jenkins off again now that the sample log files have been fixed?

**Comment by P1llus** [comment]
jenkins run the tests please

**Comment by jsoriano** [comment]
jenkins run the tests again please

**Review by kuisathaverat — CHANGES_REQUESTED** [review]
there are changes in Jenkinsfile.xml files and should not

File: dev-tools/make/mage-install.mk [hunk 1]
```diff
@@ -1,4 +1,4 @@
-MAGE_VERSION     ?= v1.10.0
+MAGE_VERSION     ?= v1.11.0
```
Comment (kuisathaverat) [inline dev-tools/make/mage-install.mk:1]: mage was update yesterday.

**Comment by hitchatwork** [comment]
@kuisathaverat - I think those changes were from rebasing from master.  The main file changes I ended up making were to:
x-pack/filebeat/module/cisco/shared/ingest/asa-ftd-pipeline.yml
x-pack/filebeat/module/cisco/asa/_meta/fields.yml
x-pack/filebeat/module/cisco/ftd/_meta/fields.yml
x-pack/filebeat/module/cisco/fields.go

The test files got updated too.

As of this writing, all tests passed locally.

If you, or @jsoriano or someone else can kick off jenkins, it should pass and should be reviewable by relevant parties.

Thanks!

**Comment by jsoriano** [comment]
@hitchatwork your branch contains several commits that are not related to your change, see https://github.com/elastic/beats/pull/22001/commits, they seem to be duplicated from master.

This is making the review a bit complicated, because it is hard to say if this would change something in an unexpected way. Could you please try to clean your branch so it includes only your changes? Feel free to create a new PR if you find it easier to copy your changes into a new branch.

Let me know if you need help with this.

**Comment by hitchatwork** [comment]
> @hitchatwork your branch contains several commits that are not related to your change, see https://github.com/elastic/beats/pull/22001/commits, they seem to be duplicated from master.
> 
> This is making the review a bit complicated, because it is hard to say if this would change something in an unexpected way. Could you please try to clean your branch so it includes only your changes? Feel free to create a new PR if you find it easier to copy your changes into a new branch.
> 
> Let me know if you need help with this.

Thanks @jsoriano - I have created a new PR #23766 to clean this up.  If you could tag/review/kick jenkins off, that'd be great.  It did pass local testing. Thanks again for your assistance!

**Comment by hitchatwork** [comment]
Opened new PR #23766 to offer cleaner review.  Closing this one out.

**Pull request closed without merging**